rollup/da_syncer/batch_queue.go

```diff
@@ -0,0 +1,103 @@
+	"context"
+	"fmt"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/da"
+)
+
+// BatchQueue is a pipeline stage that reads all batch events from DAQueue and provides only finalized batches to the next stage.
+type BatchQueue struct {
+	DAQueue                 *DAQueue
+	db                      ethdb.Database
+	lastFinalizedBatchIndex uint64
+	batches                 *common.Heap[da.Entry]
+	batchesMap              *common.ShrinkingMap[uint64, *common.HeapElement[da.Entry]]
+}
+
+func NewBatchQueue(DAQueue *DAQueue, db ethdb.Database) *BatchQueue {
+	return &BatchQueue{
+		DAQueue:                 DAQueue,
+		db:                      db,
+		lastFinalizedBatchIndex: 0,
+		batches:                 common.NewHeap[da.Entry](),
+		batchesMap:              common.NewShrinkingMap[uint64, *common.HeapElement[da.Entry]](1000),
+	}
+}
+
+// NextBatch finds next finalized batch and returns data, that was committed in that batch
+func (bq *BatchQueue) NextBatch(ctx context.Context) (da.Entry, error) {
+	if batch := bq.getFinalizedBatch(); batch != nil {
+		return batch, nil
+	}
+
+	for {
+		daEntry, err := bq.DAQueue.NextDA(ctx)
+		if err != nil {
+			return nil, err
+		}
+		switch daEntry.Type() {
+		case da.CommitBatchV0Type, da.CommitBatchWithBlobType:
+			bq.addBatch(daEntry)
+		case da.RevertBatchType:
+			bq.deleteBatch(daEntry)
+		case da.FinalizeBatchType:
+			if daEntry.BatchIndex() > bq.lastFinalizedBatchIndex {
+				bq.lastFinalizedBatchIndex = daEntry.BatchIndex()
+			}
+
+			if batch := bq.getFinalizedBatch(); batch != nil {
+				return batch, nil
+			}
+		default:
+			return nil, fmt.Errorf("unexpected type of daEntry: %T", daEntry)
+		}
+	}
+}
+
+// getFinalizedBatch returns next finalized batch if there is available
+func (bq *BatchQueue) getFinalizedBatch() da.Entry {
+	if bq.batches.Len() == 0 {
+		return nil
+	}
+
+	batch := bq.batches.Peek().Value()
+	if batch.BatchIndex() <= bq.lastFinalizedBatchIndex {
+		bq.deleteBatch(batch)
+		return batch
+	} else {
+		return nil
+	}
+}
+
+func (bq *BatchQueue) addBatch(batch da.Entry) {
+	heapElement := bq.batches.Push(batch)
+	bq.batchesMap.Set(batch.BatchIndex(), heapElement)
+}
+
+// deleteBatch deletes data committed in the batch from map, because this batch is reverted or finalized
+// updates DASyncedL1BlockNumber
+func (bq *BatchQueue) deleteBatch(batch da.Entry) {
+	batchHeapElement, exists := bq.batchesMap.Get(batch.BatchIndex())
+	if !exists {
+		return
+	}
+
+	bq.batchesMap.Delete(batch.BatchIndex())
+	bq.batches.Remove(batchHeapElement)
+
+	// we store here min height of currently loaded batches to be able to start syncing from the same place in case of restart
+	// TODO: we should store this information when the batch is done being processed to avoid inconsistencies
+	rawdb.WriteDASyncedL1BlockNumber(bq.db, batch.L1BlockNumber()-1)
```

Comment:
Any plan to adjust this TODO? I'm asking because if we ever run into such inconsistency it's hard to debug/recover. We have a similar issue with l1 message sync that we haven't been able to debug

Reply:
There is an inconsistency in the sense that we write the last processed batch here, but we haven't actually processed all blocks from this batch as they are still in the pipeline. If the node crashes or we shutdown in the meantime the DB is in an inconsistent state. This is not optimal and in a perfect world we'd already write the correct state. 
However, since the pipeline might also be started from (slightly incorrect) user input the pipeline itself needs to be able to recover to create a sequence of consecutive blocks. So whether there is an inconsistency due to the aforementioned TODO or misconfigured user input doesn't really matter here. 

Specifically, we're checking for the blocks being consecutive here before processing:
https://github.com/scroll-tech/go-ethereum/blob/ac8164f5a4190ff9e536f296195013ea7a4e3e3d/rollup/da_syncer/da_syncer.go#L30-L37

Are handling those cases with a skip in case we already processed the block:
https://github.com/scroll-tech/go-ethereum/blob/ac8164f5a4190ff9e536f296195013ea7a4e3e3d/rollup/da_syncer/syncing_pipeline.go#L189

Or reset the pipeline by moving a little bit backwards and read a previous batch from L1:
https://github.com/scroll-tech/go-ethereum/blob/ac8164f5a4190ff9e536f296195013ea7a4e3e3d/rollup/da_syncer/syncing_pipeline.go#L195
https://github.com/scroll-tech/go-ethereum/blob/ac8164f5a4190ff9e536f296195013ea7a4e3e3d/rollup/da_syncer/syncing_pipeline.go#L221

With this mechanism, in all of my testing the pipeline was able to recover from any inconsistent state I started it with.